File: packages/core/src/daemon.ts
// Resolve the pupptyeer daemon socket and ensure a daemon is running.
import net from "node:net";
import os from "node:os";
import path from "node:path";
import { spawn } from "node:child_process";
import { PupptyeerClient } from "@petersr/pupptyeer-client";

export interface DaemonOptions {
  /** Override the daemon socket path. Falls back to env / XDG / tmp. */
  socketPath?: string;
  /** Path or name of the pupptyeer binary. Falls back to env / PATH. */
  pupptyeerBin?: string;
}

/** Mirror of the daemon's socket-path resolution (see pupptyeer --help). */
export function resolveSocketPath(opts: DaemonOptions = {}): string {
  if (opts.socketPath) return opts.socketPath;
  if (process.env.PUPPTYEER_SOCK) return process.env.PUPPTYEER_SOCK;
  if (process.env.XDG_RUNTIME_DIR) {
    return path.join(process.env.XDG_RUNTIME_DIR, "pupptyeer", "daemon.sock");
  }
  const tmp = process.env.TMPDIR || "/tmp";
  return path.join(tmp, `pupptyeer-${os.userInfo().uid}`, "daemon.sock");
}

/** Path to the pupptyeer binary. Caller supplies it; else env; else PATH. */
function resolveBinary(opts: DaemonOptions): string {
  return opts.pupptyeerBin || process.env.PUPPTYEER_BIN || "pupptyeer";
}

function canConnect(sockPath: string): Promise<boolean> {
  return new Promise((resolve) => {
    const sock = net.createConnection({ path: sockPath });
    const done = (ok: boolean) => {
      sock.removeAllListeners();
      sock.destroy();
      resolve(ok);
    };
    sock.once("connect", () => done(true));
    sock.once("error", () => done(false));
    setTimeout(() => done(false), 1000);
  });
}

const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

/**
 * Connect to the pupptyeer daemon, spawning one if the socket is dead.
 * Returns a connected client.
 */
export async function connectDaemon(opts: DaemonOptions = {}): Promise<PupptyeerClient> {
  const sockPath = resolveSocketPath(opts);

  if (!(await canConnect(sockPath))) {
    const bin = resolveBinary(opts);
    console.log(`[daemon] no live socket at ${sockPath}; spawning ${bin} daemon`);
    const child = spawn(bin, ["daemon"], {
      detached: true,
      stdio: "ignore",
      env: process.env,
    });
    child.unref();

    let up = false;
    for (let i = 0; i < 50; i++) {
      if (await canConnect(sockPath)) {
        up = true;
        break;
      }
      await sleep(100);
    }
    if (!up) {
      throw new Error(
        `pupptyeer daemon did not come up at ${sockPath}. Set pupptyeerBin/PUPPTYEER_BIN or start it manually.`,
      );
    }
  }

  console.log(`[daemon] connected at ${sockPath}`);
  return PupptyeerClient.connect(sockPath);
}

File: packages/python/claude_pty_harness/daemon.py
"""Resolve the pupptyeer daemon socket and ensure a daemon is running.
Mirrors daemon.ts."""
from __future__ import annotations

import asyncio
import os
import socket
import subprocess
from dataclasses import dataclass
from typing import Optional

from ._pupptyeer import PupptyeerClient


@dataclass
class DaemonOptions:
    socket_path: Optional[str] = None
    pupptyeer_bin: Optional[str] = None


def resolve_socket_path(opts: DaemonOptions = DaemonOptions()) -> str:
    if opts.socket_path:
        return opts.socket_path
    if os.environ.get("PUPPTYEER_SOCK"):
        return os.environ["PUPPTYEER_SOCK"]
    xdg = os.environ.get("XDG_RUNTIME_DIR")
    if xdg:
        return os.path.join(xdg, "pupptyeer", "daemon.sock")
    tmp = os.environ.get("TMPDIR", "/tmp")
    return os.path.join(tmp, f"pupptyeer-{os.getuid()}", "daemon.sock")


def _resolve_binary(opts: DaemonOptions) -> str:
    return opts.pupptyeer_bin or os.environ.get("PUPPTYEER_BIN") or "pupptyeer"


def _can_connect(path: str) -> bool:
    s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    try:
        s.settimeout(1.0)
        s.connect(path)
        return True
    except OSError:
        return False
    finally:
        s.close()


async def connect_daemon(opts: DaemonOptions = DaemonOptions()) -> PupptyeerClient:
    sock = resolve_socket_path(opts)
    if not _can_connect(sock):
        binp = _resolve_binary(opts)
        print(f"[daemon] no live socket at {sock}; spawning {binp} daemon")
        subprocess.Popen(
            [binp, "daemon"],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.DEVNULL,
            start_new_session=True,
        )
        up = False
        for _ in range(50):
            if _can_connect(sock):
                up = True
                break
            await asyncio.sleep(0.1)
        if not up:
            raise RuntimeError(
                f"pupptyeer daemon did not come up at {sock}. "
                "Set pupptyeer_bin/PUPPTYEER_BIN or start it manually."
            )
    print(f"[daemon] connected at {sock}")
    return await asyncio.to_thread(PupptyeerClient.connect, sock)
